### PR TITLE
fix and test, bug correction and more test for AFD

### PR DESCRIPTION
**Pull request by Jason-Zhangxin-Chen**
Bug correction for AFD contract to make it more robust in edge cases.
It also add new tests for AFD contract, and the AFD contract test was refined as well.
More tests for fault detector are also included base on the msg aggregation.

### CODE REVIEW

File: p2p/peer_error.go
Comment (lorenzo-dev1) [inline p2p/peer_error.go:1]: Could you remove these changes from the PR? 

Once https://github.com/autonity/autonity/pull/992 is approved and merged I will just rebase the `bls-integration` branch on top of it.

File: consensus/tendermint/accountability/contracts_test.go [hunk 1]
```diff
@@ -1188,32 +1421,38 @@ func makeSigner(key blst.SecretKey) message.Signer {
...
 
-func newValidatedLightProposal(t *testing.T, height uint64, r int64, vr int64, signer message.Signer, committee types.Committee,
+func newValidatedLightProposal(height uint64, r int64, vr int64, signer message.Signer, committee types.Committee,
 	lastHeader *types.Header, block *types.Block, idx int) *message.LightProposal { //nolint
 	rawProposal := newValidatedProposalMessage(height, r, vr, signer, committee, block, idx)
-	err := rawProposal.PreValidate(lastHeader)
-	require.NoError(t, err)
-	err = rawProposal.Validate()
-	require.NoError(t, err)
+	if err := rawProposal.PreValidate(lastHeader); err != nil {
```
Comment (lorenzo-dev1) [inline consensus/tendermint/accountability/contracts_test.go:1427]: These calls to `PreValidate` and `Validate` are not necessary, all the `New*` function for consensus messages return already validated messages which are ready for use.

File: consensus/tendermint/accountability/contracts_test.go [hunk 2]
```diff
@@ -1188,32 +1421,38 @@ func makeSigner(key blst.SecretKey) message.Signer {
...
 
-func newValidatedPrecommit(t *testing.T, r int64, height uint64, v common.Hash, signer message.Signer,
+func newValidatedPrecommit(r int64, height uint64, v common.Hash, signer message.Signer,
 	s *types.CommitteeMember, cSize int, lastHeader *types.Header) *message.Precommit {
 	preCommit := message.NewPrecommit(r, height, v, signer, s, cSize)
-	err := preCommit.PreValidate(lastHeader)
-	require.NoError(t, err)
-	err = preCommit.Validate()
-	require.NoError(t, err)
+	if err := preCommit.PreValidate(lastHeader); err != nil {
```
Comment (lorenzo-dev1) [inline consensus/tendermint/accountability/contracts_test.go:1439]: Same here

File: consensus/tendermint/accountability/contracts_test.go [hunk 3]
```diff
@@ -1188,32 +1421,38 @@ func makeSigner(key blst.SecretKey) message.Signer {
...
 
-func newValidatedPrevote(t *testing.T, r int64, height uint64, v common.Hash, signer message.Signer,
+func newValidatedPrevote(r int64, height uint64, v common.Hash, signer message.Signer,
 	s *types.CommitteeMember, cSize int, lastHeader *types.Header) *message.Prevote {
 	prevote := message.NewPrevote(r, height, v, signer, s, cSize)
-	err := prevote.PreValidate(lastHeader)
-	require.NoError(t, err)
-	err = prevote.Validate()
-	require.NoError(t, err)
+	if err := prevote.PreValidate(lastHeader); err != nil {
```
Comment (lorenzo-dev1) [inline consensus/tendermint/accountability/contracts_test.go:1451]: same here

**Review by lorenzo-dev1 — CHANGES_REQUESTED** [review]
Thanks Jason!

Looks good except for a couple of minor comments.

File: p2p/peer_error.go
Comment (Jason-Zhangxin-Chen) [reply p2p/peer_error.go:1]: resolved in 68bb20784cb9458b98712f98df6949f8dcae6dde

File: consensus/tendermint/accountability/contracts_test.go (hunk 1 above)
Comment (Jason-Zhangxin-Chen) [reply consensus/tendermint/accountability/contracts_test.go]: resolved in 68bb20784cb9458b98712f98df6949f8dcae6dde

File: consensus/tendermint/accountability/contracts_test.go (hunk 2 above)
Comment (Jason-Zhangxin-Chen) [reply consensus/tendermint/accountability/contracts_test.go]: resolved in 68bb20784cb9458b98712f98df6949f8dcae6dde

File: consensus/tendermint/accountability/contracts_test.go (hunk 3 above)
Comment (Jason-Zhangxin-Chen) [reply consensus/tendermint/accountability/contracts_test.go]: resolved in 68bb20784cb9458b98712f98df6949f8dcae6dde

**Review by lorenzo-dev1 — APPROVED** [review]
LGTM
